experimental/devices/hd44780/hd44780.go

```diff
@@ -0,0 +1,261 @@
+// Copyright 2018 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+// Package hd44780 controls the Hitachi LCD display chipset HD-44780
+//
+// Datasheet
+//
+// https://www.sparkfun.com/datasheets/LCD/HD44780.pdf
+package hd44780
+
+import (
+	"fmt"
+	"time"
+
+	"periph.io/x/periph/conn"
+	"periph.io/x/periph/conn/gpio"
+)
+
+// lineTwo offset for the second line in the LCD buffer.
+const lineTwo = 0x40
+
+// Dev is the 4-bit addressing device for HD-44780
+type Dev struct {
+	// data pins
+	dataPins []gpio.PinOut
+
+	// register select pin
+	rsPin gpio.PinOut
+
+	// enable pin
+	enablePin gpio.PinOut
+}
+
+// New creates and initializes the LCD device
+//	data - references to data pins
+//	rs - rs pin
+//	e - strobe pin
+func New(data []gpio.PinOut, rs, e gpio.PinOut) (*Dev, error) {
+	if len(data) != 4 {
+		return nil, fmt.Errorf("expected 4 data pins, passed %s", len(data))
+	}
+	dev := &Dev{
+		dataPins:  data,
+		enablePin: e,
+		rsPin:     rs,
+	}
+	if err := dev.Init(); err != nil {
+		return nil, err
+	}
+	return dev, nil
+}
+
+// Init initializes the HC-44780 chipset, clears the screen buffer and moves cursor to the
+// home of screen (line 0, column 0).
+func (r *Dev) Init() error {
+	if err := r.clearBits(); err != nil {
+		return err
+	}
+
+	delayMs(15)
+
+	if err := r.rsPin.Out(gpio.Low); err != nil {
+		return err
+	}
+	if err := r.enablePin.Out(gpio.Low); err != nil {
+		return err
+	}
+
+	if err := r.bulkSendData(resetSequence, r.write4Bits); err != nil {
+		return err
+	}
+
+	if err := r.bulkSendData(initSequence, r.writeInstruction); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Halt will re-initialize the display.
+func (r *Dev) Halt() error {
+	return r.Init()
+}
+
+func (r *Dev) String() string {
+	return "HD44870, 4 bit mode"
```

Comment:
Do you think it's worth implementing 8 bits mode?

Reply:
I don't think so, it requires too much of pins. However it could be helpful to notify the end user that we use 4 bit mode for now. I don't see any advantage in 8-bit mode as for now.